arch/rv32i/src/syscall.rs

```diff
@@ -0,0 +1,379 @@
+
+          // Restore all of the app registers from what we saved. If this is the
+          // first time running the app then most of these values are
+          // irrelevant, However we do need to set the four arguments to the
+          // `_start_ function in the app. If the app has been executing then this
+          // allows the app to correctly resume.
+          mv   t0,  $2       // Save the state pointer to a specific register.
+          lw   x1,  0*4(t0)  // ra
+          lw   x2,  1*4(t0)  // sp
+          lw   x3,  2*4(t0)  // gp
+          lw   x4,  3*4(t0)  // tp
+          lw   x6,  5*4(t0)  // t1
+          lw   x7,  6*4(t0)  // t2
+          lw   x8,  7*4(t0)  // s0,fp
+          lw   x9,  8*4(t0)  // s1
+          lw   x10, 9*4(t0)  // a0
+          lw   x11, 10*4(t0) // a1
+          lw   x12, 11*4(t0) // a2
+          lw   x13, 12*4(t0) // a3
+          lw   x14, 13*4(t0) // a4
+          lw   x15, 14*4(t0) // a5
+          lw   x16, 15*4(t0) // a6
+          lw   x17, 16*4(t0) // a7
+          lw   x18, 17*4(t0) // s2
+          lw   x19, 18*4(t0) // s3
+          lw   x20, 19*4(t0) // s4
+          lw   x21, 20*4(t0) // s5
+          lw   x22, 21*4(t0) // s6
+          lw   x23, 22*4(t0) // s7
+          lw   x24, 23*4(t0) // s8
+          lw   x25, 24*4(t0) // s9
+          lw   x26, 25*4(t0) // s10
+          lw   x27, 26*4(t0) // s11
+          lw   x28, 27*4(t0) // t3
+          lw   x29, 28*4(t0) // t4
+          lw   x30, 29*4(t0) // t5
+          lw   x31, 30*4(t0) // t6
+          lw   x5,  4*4(t0)  // t0. Do last since we overwrite our pointer.
+
+          // Call mret to jump to where mepc points, switch to user mode, and
+          // start running the app.
+          mret
+
+
+
+
+          // This is where the trap handler jumps back to after the app stops
+          // executing.
+        _return_to_kernel:
+
+          // We have already stored the app registers in the trap handler. We
+          // can restore the kernel registers before resuming kernel code.
+          lw   x1,  3*4(sp)
+          lw   x3,  4*4(sp)
+          lw   x4,  5*4(sp)
+          lw   x5,  6*4(sp)
+          lw   x6,  7*4(sp)
+          lw   x7,  8*4(sp)
+          lw   x8,  9*4(sp)
+          lw   x9,  10*4(sp)
+          lw   x10, 11*4(sp)
+          lw   x11, 12*4(sp)
+          lw   x12, 13*4(sp)
+          lw   x13, 14*4(sp)
+          lw   x14, 15*4(sp)
+          lw   x15, 16*4(sp)
+          lw   x16, 17*4(sp)
+          lw   x17, 18*4(sp)
+          lw   x18, 19*4(sp)
+          lw   x19, 20*4(sp)
+          lw   x20, 21*4(sp)
+          lw   x21, 22*4(sp)
+          lw   x22, 23*4(sp)
+          lw   x23, 24*4(sp)
+          lw   x24, 25*4(sp)
+          lw   x25, 26*4(sp)
+          lw   x26, 27*4(sp)
+          lw   x27, 28*4(sp)
+          lw   x28, 29*4(sp)
+          lw   x29, 30*4(sp)
+          lw   x30, 31*4(sp)
+          lw   x31, 32*4(sp)
+
+          addi sp, sp, 33*4   // Reset kernel stack pointer
```

Comment:
There is a lot of assembly here. At this point, however, with the kernel registers restored, we should be able to switch back to rust. The reason it is all in assembly is that the assembly needs the reference to `state`, but once I pass the `&mut state` to `asm!()` Rust complains whenever I try to use it again. If someone knows how to remedy this we can probably remove some of this assembly.